--- a/drivers/mtd/nand/spi/core.c
+++ b/drivers/mtd/nand/spi/core.c
@@ -612,6 +612,34 @@ static int spinand_lock_block(struct spi
 	return spinand_write_reg_op(spinand, REG_BLOCK_LOCK, lock);
 }
 
+static int spinand_read_page_wait(struct spinand_device *spinand, u8 *s)
+{
+	unsigned long timeo = jiffies + msecs_to_jiffies(400);
+	u8 status;
+	int ret;
+
+	do {
+		ret = spinand_read_status(spinand, &status);
+		if (ret)
+			return ret;
+
+		if (status & STATUS_BUSY)
+			continue;
+
+		ret = spinand_read_status(spinand, &status);
+		if (ret)
+			return ret;
+
+		if (!(status & STATUS_BUSY))
+			break;
+
+	} while (time_before(jiffies, timeo));
+
+	*s = status;
+
+	return status & STATUS_BUSY ? -ETIMEDOUT : 0;
+}
+
 static int spinand_read_page(struct spinand_device *spinand,
 			     const struct nand_page_io_req *req)
 {
@@ -627,12 +655,19 @@ static int spinand_read_page(struct spin
 	if (ret)
 		return ret;
 
-	ret = spinand_wait(spinand,
-			   SPINAND_READ_INITIAL_DELAY_US,
-			   SPINAND_READ_POLL_DELAY_US,
-			   &status);
-	if (ret < 0)
-		return ret;
+	/* Workaround for Skyhigh */
+	if (spinand->id.data[0] == 0x01) {
+		ret = spinand_read_page_wait(spinand, &status);
+		if (ret)
+			return ret;
+	} else {
+		ret = spinand_wait(spinand,
+				   SPINAND_READ_INITIAL_DELAY_US,
+				   SPINAND_READ_POLL_DELAY_US,
+				   &status);
+		if (ret < 0)
+			return ret;
+	}
 
 	spinand_ondie_ecc_save_status(nand, status);
 
@@ -756,10 +791,17 @@ static int spinand_mtd_continuous_page_r
 		if (ret)
 			goto end_cont_read;
 
-		ret = spinand_wait(spinand, SPINAND_READ_INITIAL_DELAY_US,
-				   SPINAND_READ_POLL_DELAY_US, NULL);
-		if (ret < 0)
-			goto end_cont_read;
+		/* Workaround for Skyhigh */
+		if (spinand->id.data[0] == 0x01) {
+			ret = spinand_read_page_wait(spinand, &status);
+			if (ret)
+				goto end_cont_read;
+		} else {
+			ret = spinand_wait(spinand, SPINAND_READ_INITIAL_DELAY_US,
+					   SPINAND_READ_POLL_DELAY_US, NULL);
+			if (ret < 0)
+				goto end_cont_read;
+		}
 
 		ret = spinand_read_from_cache_op(spinand, &iter.req);
 		if (ret)
