--- drivers/mtd/nand/spi/fmsh.c 2026-01-06 02:19:22.000000000 +0530
+++ drivers/mtd/nand/spi/fmsh.c     2026-06-29 20:15:58.536130627 +0530
@@ -29,6 +29,35 @@ static SPINAND_OP_VARIANTS(update_cache_
                SPINAND_PROG_LOAD_1S_1S_4S_OP(false, 0, NULL, 0),
                SPINAND_PROG_LOAD_1S_1S_1S_OP(false, 0, NULL, 0));

+static int fm25g01b_ooblayout_ecc(struct mtd_info *mtd, int section,
+                                 struct mtd_oob_region *region)
+{
+       if (section)
+               return -ERANGE;
+
+       region->offset = 64;
+       region->length = 64;
+
+       return 0;
+}
+
+static int fm25g01b_ooblayout_free(struct mtd_info *mtd, int section,
+                                  struct mtd_oob_region *region)
+{
+       if (section)
+               return -ERANGE;
+
+       region->offset = 2;
+       region->length = 62;
+
+       return 0;
+}
+
+static const struct mtd_ooblayout_ops fm25g01b_ooblayout = {
+       .ecc = fm25g01b_ooblayout_ecc,
+       .rfree = fm25g01b_ooblayout_free,
+};
+
 static int fm25s01a_ooblayout_ecc(struct mtd_info *mtd, int section,
                                  struct mtd_oob_region *region)
 {
@@ -53,6 +82,24 @@ static const struct mtd_ooblayout_ops fm
 };

 static const struct spinand_info fmsh_spinand_table[] = {
+       SPINAND_INFO("FM25G01B",
+                    SPINAND_ID(SPINAND_READID_METHOD_OPCODE_DUMMY, 0xd1),
+                    NAND_MEMORG(1, 2048, 128, 64, 1024, 21, 1, 1, 1),
+                    NAND_ECCREQ(8, 528),
+                    SPINAND_INFO_OP_VARIANTS(&read_cache_variants,
+                                             &write_cache_variants,
+                                             &update_cache_variants),
+                    0,
+                    SPINAND_ECCINFO(&fm25g01b_ooblayout, NULL)),
+       SPINAND_INFO("FM25G02B",
+                    SPINAND_ID(SPINAND_READID_METHOD_OPCODE_DUMMY, 0xd2),
+                    NAND_MEMORG(1, 2048, 128, 64, 2048, 41, 1, 1, 1),
+                    NAND_ECCREQ(8, 528),
+                    SPINAND_INFO_OP_VARIANTS(&read_cache_variants,
+                                             &write_cache_variants,
+                                             &update_cache_variants),
+                    0,
+                    SPINAND_ECCINFO(&fm25g01b_ooblayout, NULL)),
        SPINAND_INFO("FM25S01A",
                     SPINAND_ID(SPINAND_READID_METHOD_OPCODE_DUMMY, 0xE4),
                     NAND_MEMORG(1, 2048, 64, 64, 1024, 20, 1, 1, 1),